File: Javascript(一个前端工程师的挣扎）/Two Pointers/Common Problems/56.merge_intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
    if (intervals === null || intervals.length === 0) return intervals;
    intervals.sort(function(a, b) {
        return a[0] - b[0];
    })

    let result = [];
    let last = intervals[0];

    for (let i = 1; i < intervals.length; i++) {
        let cur = intervals[i];

        if (cur[0] <= last[1]) {
            last[1] = Math.max(cur[1], last[1]); // 这里有个大BUG,因为不是第二个INTERVAL的结尾就一定比第一个的结尾大
        } else {
            res.push(last);
            last = cur;
        }
    }
    res.push(last); //这个值漏了，写完代码最后再自己过一遍

    return res;
};

File: Java/56.merge_intervals.java
// Date: 11-21-2019
// Tag: Heap
// Input: [[1,3],[2,6],[8,10],[15,18]]
//        Output: [[1,6],[8,10],[15,18]]
//        Explanation: Since intervals [1,3] and [2,6] overlaps, merge them into [1,6].
class Solution {
    public int[][] merge(int[][] intervals) {
        if (intervals == null || intervals.length == 0) return intervals;

        Arrays.sort(intervals, new Comparator<int[]>() {
            @Override
            public int compare(int[] a, int[] b) {
                if (a[0] == b[0]) return a[1] - b[1];
                return a[0] - b[0];
            }
        });

        // After sorting the list, it is sorted by start time, then end time

        List<int[]> res = new ArrayList<>();
        int[] newInterval = intervals[0];
        res.add(newInterval);

        for (int[] interval : intervals) {
            if (interval[0] <= newInterval[1]) {
                newInterval[1] = Math.max(newInterval[1], interval[1]);
            } else {
                newInterval = interval;
                res.add(newInterval);
            }
        }

        return res.toArray(new int[res.size()][]);
    }
}